compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala

```diff
@@ -95,58 +95,79 @@ object DesugarEnums {
-   *       {
-   *         case ex$: NoSuchElementException =>
-   *           throw new IllegalArgumentException("enum case not found: " + $name)
-   *       }
+   *     val nameHash = if $name == null then 0 else $name.hashCode
+   *     var ordinal = -1
+   *     nameHash match {
+   *       case 100054 =>
+   *         if "e_0" equals $name then
+   *           ordinal = 0
+   *       ...
+   *       case 100116 =>
+   *         if "e_n" equals $name then
+   *           ordinal = n
+   *       case _ =>
+   *     }
+   *     ordinal match {
+   *       case 0 => e_0
+   *       ...
+   *       case n => e_n
+   *       case _ => throw new IllegalArgumentException("case not found: " + $name)
+   *     }
    */
   private def enumScaffolding(enumCases: List[(Int, TermName)])(using Context): List[Tree] = {
     import dotty.tools.dotc.transform.SymUtils.rawTypeRef
+
+    def const(arg: String | Int | Null | Unit) = Literal(Constant(arg))
+
     val rawEnumClassRef = rawRef(enumClass.typeRef)
     extension (tpe: NamedType) def ofRawEnum = AppliedTypeTree(ref(tpe), rawEnumClassRef)
 
     val privateValuesDef =
       ValDef(nme.DOLLAR_VALUES, TypeTree(), JavaSeqLiteral(enumCases.map((_, name) => Ident(name)), rawEnumClassRef))
         .withFlags(Private | Synthetic)
 
-    val privateReverseValuesDef =
-      val wrapped = Apply(Select(ref(defn.ScalaRuntimeModule.termRef), nme.wrapRefArray), Ident(nme.DOLLAR_VALUES))
-      val mapper =
-        val paramName = nme.syntheticParamName(0)
-        val paramDef = param(paramName, rawEnumClassRef)
-        Function(paramDef :: Nil, Tuple(Select(Ident(paramName), nme.enumLabel) :: Ident(paramName) :: Nil))
-      val mapBody = Select(Apply(Select(wrapped, nme.map), mapper), nme.toMap)
-      val annot = New(ref(defn.ThreadUnsafeAnnot.typeRef), Nil).withSpan(ctx.tree.span)
-      ValDef(nme.DOLLAR_VALUES_REVERSE, TypeTree(), mapBody)
-          .withFlags(Private | Synthetic | Lazy).withAnnotations(annot :: Nil)
-
     val valuesDef =
       DefDef(nme.values, Nil, Nil, defn.ArrayType.ofRawEnum, valuesDot(nme.clone_))
         .withFlags(Synthetic)
 
-    val valuesOfExnMessage = Apply(
-      Select(Literal(Constant("enum case not found: ")), nme.PLUS), Ident(nme.nameDollar))
-
-    val valuesOfBody = Try(
-      expr = Apply(Ident(nme.DOLLAR_VALUES_REVERSE), Ident(nme.nameDollar) :: Nil),
-      cases = CaseDef(
-        pat = Typed(Ident(nme.DEFAULT_EXCEPTION_NAME), TypeTree(defn.NoSuchElementExceptionType)),
-        guard = EmptyTree,
-        body = Throw(New(TypeTree(defn.IllegalArgumentExceptionType), List(valuesOfExnMessage :: Nil)))
-      ) :: Nil,
-      finalizer = EmptyTree
-    )
+    val valuesOfBody: Tree =
+      val nameHash = nme.syntheticParamName(0)
+      val ordinal = nme.syntheticParamName(1)
+      val nameHashDef =
+        ValDef(nameHash, TypeTree(defn.IntType), If(
+          cond = Apply(Select(Ident(nme.nameDollar), nme.eq), const(null)),
```

Comment:
Any reason to not just let it crash with an NPE if the user passes null?

Reply:
not really, just would be the same as the behaviour if it was a match expression